Properly compute resource folder when linking statically

drexin opened the pull request:
- deduplicate the logic to compute the resource folder
- install headers and module files in shared and static resource folders
- forward -static flag when calling swiftc with -print-target-info

drexin commented:
@swift-ci test

compnerd commented on include/swift/Driver/ToolChain.h:
```diff
@@ -205,7 +205,7 @@ class ToolChain {
   /// Get the resource dir link path, which is platform-specific and found
   /// relative to the compiler.
-  void getResourceDirPath(SmallVectorImpl<char> &runtimeLibPath,
+  void getResourceDirPath(SmallString<128> &runtimeLibPath,
```
Why this change?  The `SmallVectorImpl` is the stack allocation size erased base type for `SmallString`.  This allows for different sizes to be used.

drexin replied:
reverted

compnerd commented on include/swift/Frontend/Frontend.h:
```diff
@@ -128,7 +128,8 @@ class CompilerInvocation {
                      *ConfigurationFileBuffers = nullptr,
-                 StringRef workingDirectory = {});
+                 StringRef workingDirectory = {},
+                 StringRef mainExecutablePath = {});
```
Indentation seems off

compnerd commented on include/swift/Frontend/Frontend.h:
```diff
@@ -214,8 +215,16 @@ class CompilerInvocation {
   static void computeRuntimeResourcePathFromExecutablePath(
       StringRef mainExecutablePath,
+      bool shared,
       llvm::SmallString<128> &runtimeResourcePath);
```
`clang-format` please

compnerd commented on lib/Driver/Driver.cpp:
```diff
@@ -2255,6 +2255,14 @@ bool Driver::handleImmediateArgs(const ArgList &Args, const ToolChain &TC) {
+                     options::OPT_no_static_stdlib, false) ||
+        Args.hasArg(options::OPT_static)) {
+      commandLine.push_back("-static");
+    }
```
I thought `-static-stdlib` was for statically linking the standard library but creating a dynamic executable/library.  Is that not the case?  If not, this makes sense, otherwise, this seems wrong because `-static` will will always do static linking.

drexin replied:
The linking is handled by the driver and `OPT_static` is not used in any form in the frontend currently. I'm happy to change it to something else, but the frontend does not seem to know `OPT_static_stdlib` or `OPT_static_executable`.

compnerd replied:
Hmm, I suppose that my discomfort comes from the difference in what `-static` means when you are using `clang` and `swift`.  It just seems very easy to confuse someone (as opposed to many of the other arguments that have different spellings).

`-static`, `-static-executable` both imply creating a static binary.  `-static-stdlib` says that statically link the standard library but still still generate a dynamically linked binary.  As long `-static` is not forwarded from the frontend to the driver it will be okay I suppose, just confusing (non-Darwin targets use the linker driver (`clang`) to invoke the linker which will process `-static` as generating a static binary).  If you want to still go with this, I think that this deserves a comment explaining what is going on here and how this fits together.

drexin replied:
I added a new flag to the frontend `-use-static-resource-dir`

compnerd reviewed (commented):
Could you please add a test for the driver path computation?

drexin committed:
Properly compute resource folder when linking statically

- deduplicate the logic to compute the resource folder
- install headers and module files in shared and static resource folders
- forward -static flag when calling swiftc with -print-target-info

compnerd commented on include/swift/Frontend/Frontend.h:
```diff
@@ -213,9 +214,14 @@ class CompilerInvocation {
       llvm::SmallString<128> &runtimeResourcePath);
 
+  /// Appends `lib/swift[_static]` to the given path
+  static void appendLibDir(llvm::SmallString<128> &path, bool shared);
```
I would say that `SmallVectorImpl<char>` is better than the hardcoded stack size.

compnerd commented on include/swift/Frontend/Frontend.h:
```diff
@@ -213,9 +214,14 @@ class CompilerInvocation {
+  /// Appends `lib/swift[_static]` to the given path
+  static void appendLibDir(llvm::SmallString<128> &path, bool shared);
+
+  void computeRuntimeResourcePath(llvm::SmallString<128> &runtimeResourcePath);
```
Similar

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test macos

drexin commented:
@swift-ci test macos

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test windows

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test windows

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test macos

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test macos

drexin commented:
@swift-ci test windows

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test windows

drexin commented:
@swift-ci test macos

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test linux

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test

drexin commented:
@swift-ci test windows

drexin commented:
@swift-ci test windows

compnerd commented on include/swift/Frontend/Frontend.h:
```diff
@@ -213,8 +214,11 @@ class CompilerInvocation {
+      llvm::SmallVectorImpl<char> &runtimeResourcePath);
+
+  /// Appends `lib/swift[_static]` to the given path
+  static void appendLibDir(llvm::SmallVectorImpl<char> &path, bool shared);
```
Would be nice to make this `appendSwiftLibDir` since `libdir` in most contexts here tends to refer to `lib` vs `lib32` or `lib64`.

compnerd commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -283,6 +283,8 @@ class FrontendOptions {
   /// -dump-scope-maps.
   SmallVector<std::pair<unsigned, unsigned>, 2> DumpScopeMapLocations;
 
+  bool UseSharedResourceFolder = true;
```
Doc comment please (its obvious in the context of the change, but less so when you just have a type to work with).

drexin commented:
@swift-ci smoke test

drexin commented:
@swift-ci smoke test

drexin commented:
@swift-ci smoke test windows

drexin commented:
@swift-ci smoke test windows

DougGregor commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1266,24 +1274,18 @@ void ToolChain::getClangLibraryPath(const ArgList &Args,
     resourceDirPath.append(value.begin(), value.end());
-    llvm::sys::path::append(resourceDirPath, "usr", "lib",
-                            shared ? "swift" : "swift_static");
+    llvm::sys::path::append(resourceDirPath, "usr");
```
Why are we adding `<resource-dir>/usr` now?

drexin replied:
We are using the SDK root here as opposed to the path relative to the executable. In the executable case we get something like `$PATH/bin/swift` and remove the `bin/swiftc` part from it and append `lib/swift[_static]`. The SDK path starts one level above that, so we have to add `usr` only in this case.
